CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,475 @@
+---
+cip: <to be assigned>
+title: Extension to ODIS for Password Hashing
+author: Victor Graf <victor@clabs.co>, Cody Born <cody@clabs.co>, Alec Schaefer <alec@clabs.co>
+discussions-to: https://github.com/celo-org/celo-proposals/issues/233
+status: Draft
+type: Standards Track
+category: 3
+created: 2021-06-15
+<!--
+  TODO(victor): Once a CIP for the cryptography is ready, add that here as a dependency.
+  requires:
+-->
+license: Apache 2.0
+---
+
+### Terminology
+
+<!-- TODO(victor): Actually go through and add these keywords where appropriate -->
+The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to be interpreted as
+described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html).
+
+## Simple Summary
+
+An extension to the ODIS phone number privacy service to support granular rate limiting with the
+goal of hardening passwords for use as encryption and authorization keys.
+
+## Abstract
+
+<!-- TODO(victor): Once we have a roughly agreed upon proposal for the pOPRF, link it here -->
+Given the required cryptographic functionality (to be discussed in more detail in a future
+proposal), this proposal describes an interface to be exposed by the ODIS service allowing the user
+to specify a domain string (i.e. a salt value) with rate-limiting rules to be combined with their
+blinded input (e.g.  phone number or password).
+
+Central to the proposed interface are domains. A domain is a struct which specifies rate limiting
+information and uniquely identifies the context of the secret to be hashed. As an example, a domain
+for hashing an account password might specify the username (context) and a cap of 10 password
+attempts (rate-limiting parameter). These would be combined with their password to form the final
+output.
+
+Queries with distinct domain specifiers will receive uncorrelated output. For example, output from
```

Comment:
Is this accomplished by using separate peppers (or salts) based on the domain? As in each ODIS signer holds a set of distinct keys (one for each existing domain) and the combiner generates the pepper by using the domain-specific keys from a threshold number of signers?

Reply:
I leave this a bit vauge in the abstract, simply stating that the domain is combined with the message, but try to calrify later. Essentially though, the domain acts as a salt, and salts are often thought of as "function selectors" when working with PRFs. (i.e. You can define from function $f(s || x)$ an equivalent function $f_s(x)$ where s is the salt and x is the message.)

Reply:
Does it become clear after reading the rest of the proposal. If not, I'll try to fix that.